lucille-core/src/main/java/com/kmwllc/lucille/connector/storageclient/S3StorageClient.java

```diff
@@ -65,108 +58,127 @@ public void shutdown() throws IOException {
         throw new IOException("Error occurred closing S3Client", e);
       }
     }
-
-    // clear all FileHandlers if any
-    clearFileHandlers();
   }
 
   @Override
-  public void traverse(Publisher publisher) throws Exception {
-    ListObjectsV2Request request = ListObjectsV2Request.builder().bucket(bucketOrContainerName).prefix(startingDirectory).maxKeys(maxNumOfPages).build();
-    ListObjectsV2Iterable response = s3.listObjectsV2Paginator(request);
-    response.stream()
-        .forEachOrdered(resp -> {
-          resp.contents().forEach(obj -> {
-            if (isValid(obj)) {
-              String fullPathStr = getFullPath(obj);
-              String fileExtension = FilenameUtils.getExtension(fullPathStr);
-              tryProcessAndPublishFile(publisher, fullPathStr, fileExtension, new FileReference(obj));
-            }
+  public void traverse(Publisher publisher, TraversalParams params) throws Exception {
+    try {
+      initializeFileHandlers(params);
+
+      ListObjectsV2Request request = ListObjectsV2Request.builder()
+          .bucket(getBucketOrContainerName(params))
+          .prefix(getStartingDirectory(params))
+          .maxKeys(maxNumOfPages).build();
+      ListObjectsV2Iterable response = s3.listObjectsV2Paginator(request);
+      response.stream()
+          .forEachOrdered(resp -> {
+            resp.contents().forEach(obj -> {
+              if (isValid(obj, params)) {
+                String fullPathStr = getFullPath(obj, params);
+                String fileExtension = FilenameUtils.getExtension(fullPathStr);
+                tryProcessAndPublishFile(publisher, fullPathStr, fileExtension, new FileReference(obj), params);
+              }
+            });
           });
-        });
+    } finally {
+      clearFileHandlers();
+    }
+  }
+
+  @Override
+  public InputStream getFileContentStream(URI uri) throws IOException {
```

Comment:
should confirm that the storage client has been initialized and not closed/shutdown